test/git_repository_test.go

```diff
@@ -176,6 +176,13 @@ func TestGitRepository_CreateFeatureBranch(t *testing.T) {
 	assert.True(t, repo.Configuration(true).IsFeatureBranch("f1"))
 }
 
+func TestGitRepository_CreateFeatureBranchNoParent(t *testing.T) {
+	repo := createTestGitTownRepo(t)
+	err := repo.CreateFeatureBranchNoParent("f2")
+	assert.Nil(t, err)
+	assert.True(t, repo.Configuration(true).IsFeatureBranch("f1"))
+}
+
```

Comment:
How do I run these tests? The "f2" branch was a typo intended to cause a test failure but it snuck through because the test never failed and I got distracted by something else.

Reply:
`make unit` runs the go unit tests

Reply:
What @charlierudolph said, and you can also run `make u` to only run the unit tests that have changed.

Reply:
I was running `make tests`, which does include running the unittests. It appears that this test case is passing when it should not be. I made the following change and now the test fails:

```
@@ -179,7 +179,7 @@ func TestGitRepository_CreateFeatureBranch(t *testing.T) {
 func TestGitRepository_CreateFeatureBranchNoParent(t *testing.T) {
        repo := createTestGitTownRepo(t)
        err := repo.CreateFeatureBranchNoParent("f2")
-       assert.Nil(t, err)
+       assert.NotNil(t, err)
        assert.True(t, repo.Configuration(true).IsFeatureBranch("f1"))
 }
```

So why does `IsFeatureBranch` pass when we didn't create an `f1` feature branch? It turns out that `IsFeatureBranch` does not check for the existence of the branch:

```
// IsFeatureBranch indicates whether the branch with the given name is
// a feature branch.
func (c *Configuration) IsFeatureBranch(branchName string) bool {
        return !c.IsMainBranch(branchName) && !c.IsPerennialBranch(branchName)
}
```

Is this something that you want to fix? It doesn't look trivial to plumb through, but maybe I'm missing something.

Reply:
So, I made a small change to `IsFeatureBranch` to check for a local branch with the given name, but the causes other tests to fail:

```
 func (c *Configuration) IsFeatureBranch(branchName string) bool {
+       if _, err := c.shell.Run("git", "rev-parse", "--verify", branchName); err != nil {
+               return false
+       }
        return !c.IsMainBranch(branchName) && !c.IsPerennialBranch(branchName)
 }
```

I will fix the simple typo but it appears that there is room for improvement here.

Reply:
Okay, last attempt that I tried still doesn't fix the simple typo issue. But at least it verifies that the parent branches exist or don't exist as desired:

```
diff --git a/test/git_repository_test.go b/test/git_repository_test.go
index 55759354..a5844232 100644
--- a/test/git_repository_test.go
+++ b/test/git_repository_test.go
@@ -174,13 +174,15 @@ func TestGitRepository_CreateFeatureBranch(t *testing.T) {
        err := repo.CreateFeatureBranch("f1")
        assert.Nil(t, err)
        assert.True(t, repo.Configuration(true).IsFeatureBranch("f1"))
+       assert.Equal(t, []string{"main"}, repo.Configuration(true).GetAncestorBranches("f1"))
 }
 
 func TestGitRepository_CreateFeatureBranchNoParent(t *testing.T) {
        repo := createTestGitTownRepo(t)
-       err := repo.CreateFeatureBranchNoParent("f2")
+       err := repo.CreateFeatureBranchNoParent("f1")
        assert.Nil(t, err)
        assert.True(t, repo.Configuration(true).IsFeatureBranch("f1"))
+       assert.Equal(t, []string(nil), repo.Configuration(true).GetAncestorBranches("f1"))
 }
 
 func TestGitRepository_CreateFile(t *testing.T) {
```

Reply:
Thanks for looking into this, @cirego! These commands really only do what their name says. In this case, the command name is `IsFeatureBranch`, meaning "would a branch with this name be considered a feature branch or a perennial/main branch?". It doesn't imply "does this branch exist?". The second question would be answered by a `HasBranch` method. Such a method doesn't exist yet because we haven't had a need for it. So far, the code always knows which branches exist. If anybody needs it, we are happy to add it.

The logic behind this (arguably slightly confusing) behavior is that running Git commands is a relatively expensive operation: we have to open a subshell, run Git, which itself has to perform a number of operations on the file system like loading the Git configuration and the branch list. We want to make it explicit when a Git command runs. This is arguably a bit more cumbersome, but has allowed us to strip many unnecessary calls to Git, which helps keeping Git Town and especially the test suite that calls it many thousands of times fast (i.e. finishing within just a few minutes). In my experience, this speed has saved way more headaches than the need to write a few more checks.  😃